Rewrite convergent evolution section as temporal narrative (#99)
myyoda/principles-paper: main.tex

The FAIR principles\cite{wilkinson_fair_2016} and FAIR4RS guidelines\cite{barker_introducing_2022} provide structured guidance for making digital objects findable, accessible, interoperable, and reusable.
The Workflow Community Initiative (WCI-FW)\cite{wilkinson_applying_2025} mapped these principles onto computational workflows but explicitly chose not to define new ones, instead treating workflows as hybrid data/software objects.

> FAIR addresses how digital objects should be described so they can be found, accessed, and reused; STAMPED addresses a complementary question: how a research object should be organized and managed so it can actually be re-executed, extended, and redistributed.

Reviewer: Similar to the follow-up purge of em-dashes, the sentence structure of semi-colon into colon is just so awkward...

TODO in next PR across the manuscript

Author: i think ';' is just there to be visually distinctive from a preceding ',' listing of FAIR features.  But I agree that here we could even simplify to have 2 smaller sentences:

 ```suggestion
FAIR addresses how digital objects should be described so they can be found, accessed, and reused.
STAMPED addresses a complementary question: how a research object should be organized and managed so it can actually be re-executed, extended, and redistributed.
```

Reviewer: Yes, better. We can either do this here or in a dedicated follow-up, rewording all such patterns across the manuscript - which do you prefer?

Author: I will do here for this one while we see it ;)

The FAIR principles\cite{wilkinson_fair_2016} and FAIR4RS guidelines\cite{barker_introducing_2022} provide structured guidance for making digital objects findable, accessible, interoperable, and reusable.
The Workflow Community Initiative (WCI-FW)\cite{wilkinson_applying_2025} mapped these principles onto computational workflows but explicitly chose not to define new ones, instead treating workflows as hybrid data/software objects.

> FAIR addresses how digital objects should be described so they can be found, accessed, and reused.

Reviewer 2: We are now duplicating line 38. This chunk needs to incorporate 38/39. 

not a blocker though, this adds helpful stuff, we could potentially merge and fixup later

Reviewer 2: removed the duplication, I think it still flows alright.

These adoptions demonstrate that STAMPED properties solve practical organizational challenges across scales, from individual studies to community-wide infrastructure.

> \subsection{Convergent evolution of empirical scientific practice}

Reviewer: ```suggestion
\subsection{Convergent evolution of empirical scientific practices}
```

Should be plural? Also a bit lengthy for a title, any way to shorten? What does the word 'empirical' bring to this?

Author: I think I chose singular on purpose: we have 'properties' and 'principles' already.  Here I use "practice" as a process or a complete work flow, not a list of "practices" as individual steps or properties/etc.

"empirical" also was on purpose to restrict our "evolution" to practice(s) relating to data-driven sciences, as opposed to some other less such, of which we do not talk/care here at all, and which could have their own challenges which do not relate to computation, data, etc.

Reviewer: I do not equate 'empirical' science with 'data-driven'

There are many wet-lab bio people, even today, that do lots of empirical science and never digitize a single item outside of a number in a spreadsheet

> I think I chose singular on purpose: we have 'properties' and 'principles' already. Here I use "practice" as a process or a complete work flow, not a list of "practices" as individual steps or properties/etc.

Interesting, but when I read the text below it is indeed about describing all the individual practices, not an introduction to a case study of a scientist engaging in scientific practice as an idea

```suggestion
\subsection{Convergent evolution of data-driven scientific practices}
```

Author: ok, let's try going with plural, indeed "data-driven" would be more specific

The challenges that motivate STAMPED are not unique to any single domain.
Over three decades, communities spanning geophysics, statistics, genomics, neuroimaging, and machine learning independently converged on strikingly similar organizational responses to the same recurring problems: fragmented provenance, irreproducible computations, and ad-hoc project layouts.

> What follows is not a comprehensive review but a tracing of that convergence, organized chronologically, to show that the properties STAMPED formalizes reflect durable responses to recurrent challenges rather than arbitrary design choices.

Reviewer: ```suggestion
What follows is not a comprehensive review but a tracing of that convergence, organized chronologically, to show that the properties STAMPED formalizes reflect durable responses to recurring challenges.
```

"recurrent" can be a formal phrase meaning different things depending on subfields (including the ones mentioned in the previous sentence)

I don't think the last part I removed added anything, and sounded a little odd

The earliest calls for reproducibility-as-organization appeared in the 1990s.

> Claerbout and Karrenbach\cite{claerbout_electronic_1992} demonstrated that coupling electronic documents with Makefiles could make geophysical analyses self-contained, tracked, and re-executable---anticipating Self-containment~(S), Tracking~(T), and Actionability~(A).

Reviewer: Interesting to note here - the coupling of "re-executable" with "Actionability" really does make the Dominos (as an antithetical example) metaphor more apt; a stack of Dominos can't be easily re-executed like a Rube-Goldberg machine can

Author: I might need you to elaborate on what you mean here to grasp the idea.

Reviewer: Seems GitHub does not like embedding GIFs...

Here is a short video of a [stack of dominoes falling](https://www.google.com/search?sa=X&sca_esv=4e0bfa84728bf4dd&sxsrf=ANbL-n7u58WAifyNHrjESBVptbETF80JEw:1774573550980&udm=39&fbs=ADc_l-aN0CWEZBOHjofHoaMMDiKpmAsnXCN5UBx17opt8eaTX5MJRoosnbembaWTjeNSquIxro2mrW6zffXrbXZY-opPcMmGGbjhze_llTX0fE0zrjJdzPBETbunNfZMofxWtZnY18AboXI639Z6RvkPaeAvozBwa7oQYZhG9Iy02zz53Op8o0YiZa2t_YWTUvg7My1a3u8HRXh0rKWTC6vBB3MciadYCg&q=video+of+dominoes+falling&ved=2ahUKEwio64jR8b6TAxXuF2IAHRsbEWsQs6gLegQIDRAB&biw=1512&bih=738&dpr=2#fpstate=ive&ip=1&vld=cid:22766901,vid:0Mw0NstPGfE,st:0)

The 'Actionability' is to make the last domino fall ('output') by initiating the input action (flicking the first domino) 

But it cannot be re-executed easily, if at all - even re-stacking the dominos will result in minor random differences in placements and could introduce new failure points at each point of internal interaction

Whereas my GIF of the Rube-Goldberg machine from my slides shows many balls being passed as input, all the cogs and wheels continuing to turn, and delivering the ball inputs to the output ball catcher - very re-executable

Reviewer 2: I like the analogy, but I still dont think "Rube goldberg" is the right counter to dominos, since they almost never set themselves up to run again. What about a bowling alley? The machine sets up the initial state, things happen, and the state is reset and ready to run again, while recording the provenance/score persistently.

Reviewer: > I still dont think "Rube goldberg" is the right counter to dominos, since they almost never set themselves up to run again. 

What doesn't set themselves to run again? Most 'good' Rube Golderberg machines should:

https://www.reddit.com/r/MotionDesign/comments/1831p72/rube_goldberg_animation/

 I guess it is not a prescriptive definition; you could have a non-perpetual one

Noble\cite{noble_quick_2009} provided the first widely adopted project-organization guide for computational biology, offering concrete directory-structure conventions that addressed Self-containment~(S) and Modularity~(M).

A wave of normative frameworks followed in the 2010s.
Stodden\cite{stodden_scientific_2010} articulated legal and social dimensions of sharing computational research, raising concerns central to Distributability~(D) and Tracking~(T).
Sandve et~al.\cite{sandve_ten_2013} distilled ten rules for reproducible computation---track versions, record workflows, archive environments---spanning Tracking~(T), Actionability~(A), and Self-containment~(S).
The FAIR principles\cite{wilkinson_fair_2016} established findability, accessibility, interoperability, and reusability as community norms for digital objects, later extended to research software by FAIR4RS\cite{barker_introducing_2022} and to workflows by the Workflow Community Initiative\cite{wilkinson_applying_2025}.
Wilson et~al.\cite{wilson_good_2017} codified ``Good Enough Practices'' that operationalized many of these ideals for working scientists---Actionability~(A), Tracking~(T), and Self-containment~(S).

> The YODA principles\cite{hanke_yoda_2018} combined these insights into organizational rules from which STAMPED directly descends.

Reviewer: This is back to my complaint about referring to current references of YODA as 'principles' rather than 'practices' in this first formally published (hopefully) reference to them

When listed in the same paragraph as FAIR, the difference is a stark contrast; if we call FAIR 'principles' (which they certainly are) then we cannot say the bullet points of the OHBM poster are equivalently 'principles' and are instead 'practices' or 'conventions'. They are also not stated as rules (no clear definitions or boundaries to achieve or violate the statements, which also read more as suggestions)

```suggestion
The FAIR principles\cite{wilkinson_fair_2016} established findability, accessibility, interoperability, and reusability as community norms for digital objects, later extended to research software by FAIR4RS\cite{barker_introducing_2022} and to workflows by the Workflow Community Initiative\cite{wilkinson_applying_2025}.
Wilson et~al.\cite{wilson_good_2017} codified ``Good Enough Practices'' that operationalized many of these ideals for working scientists---Actionability~(A), Tracking~(T), and Self-containment~(S).
The YODA conventions\cite{hanke_yoda_2018} combined these insights into organizational practices from which STAMPED directly descends.
```

Author: I like to relate to "practices" as we mention them...  but again, I would keep them principles as that's how we defined and referenced to them, hence let's agree to agree on 

```suggestion
The FAIR principles\cite{wilkinson_fair_2016} established findability, accessibility, interoperability, and reusability as community norms for digital objects, later extended to research software by FAIR4RS\cite{barker_introducing_2022} and to workflows by the Workflow Community Initiative\cite{wilkinson_applying_2025}.
Wilson et~al.\cite{wilson_good_2017} codified ``Good Enough Practices'' that operationalized many of these ideals for working scientists---Actionability~(A), Tracking~(T), and Self-containment~(S).
The YODA principles\cite{hanke_yoda_2018} combined these insights into organizational practices from which STAMPED directly descends.
```

> The YODA principles\cite{hanke_yoda_2018} combined these insights into organizational rules from which STAMPED directly descends.

Reviewer: "from which STAMPED directly descends" this is a good statement and need to check where/if it occurs elsewhere

the unfortunate thing is as we discussed above, STAMPED is effectively YODA + VAMP, but how do we refer to VAMP since they never even released a preprint?

Author: I cited Michael's talk where he presented it

Reviewer: Can you cite it here then and add mention of VAMP? To give better 'provenance' to our ideas

Reviewer 2: hanke_what_2023 is now in references.bib after merge from main and regenerate, but should probably be included here before resolving this comment

Author: we have the talk ref'ed now:

```
❯ git grep hanke_what_2023
main.tex:Building on YODA and this emphasis on actionability, the VAMP formulation\cite{hanke_what_2023} (Versioned, Actionable, Modular, Portable) made Actionability~(A), Portability~(P), and Modularity~(M) explicit organizational requirements, and introduced Ephemerality~(E) as validation for Portability~(P)---directly prefiguring STAMPED.
references.bib:@misc{hanke_what_2023,

```

Notably, the WCI-FW framework\cite{wilkinson_applying_2025} extended FAIR to computational workflows but explicitly declined to define new principles, leaving the operational gap---how a research object should be structured and managed---that STAMPED fills.

Independent of these normative efforts, a parallel ``Git for data'' convergence occurred in tooling.

> git-annex (2010), Git~LFS (2015), DVC (2017), Pachyderm (2014), and Icechunk (2024) all independently arrived at content-addressed identification of large files, logical separation of metadata from bulk storage, and derivation records linking inputs to outputs.

Reviewer: So, what is the story with Git LFS historically? Did it start as a separate service / plugin and then get absorbed into GitHub (or is the GitHub LFS an entirely separate service)?

I guess I'm not clear on that history, but IDK if it needs any expansion to describe or if we just assume interested readers to follow references to find out more

Author: claude probably would tell it better than I would ;) IIRC it was "git-media" and then coined into "git-lfs" and then indeed adopted by github . gitlab at some point had git-annex and dropped it, likely has git-lfs too now.

Author: overall -- yeah, I would not aim for us to provide neither introduction to all of them, nor comparison. References should IMHO be sufficiently enough.  But we might want to add quilt data https://github.com/quiltdata/quilt -- albeit AWS specific, it was out there for a while and always accented on versioning (IIRC with accent on tabular data)

Author: oh oh -- there also was that DAT -- https://dat-ecosystem.org/, which I believe was a precursor to ipfs (or just from the same Juan...) 

but well -- again, we are not overview and I think we cover already most known IMHO, so we could leave those for reviewers etc to identify

> git-annex (2010), Git~LFS (2015), DVC (2017), Pachyderm (2014), and Icechunk (2024) all independently arrived at content-addressed identification of large files, logical separation of metadata from bulk storage, and derivation records linking inputs to outputs.

Reviewer: Is Icechunk the publication standing in for Zarr? Otherwise Zarr does many of the things you mention (content-addressed storage, granted you might need certain filters enabled; separation of metadata from bulk; I guess it does not support links but are you sure Icechunk supports links?)

Author: Zarr here is IMHO not really 'important' even if talking about Icechunk, so I would not go into the realm of file/directory formats here.

Despite differing architectures---some cloud-native, some local-first, some centralized, some federated---each converged on the same core insight: that version-control semantics proven for code could and should extend to data.
This convergence is evidence not of influence but of shared need, and maps most directly onto Tracking~(T).

A complementary convergence emerged in platforms and metadata standards.

> Galaxy\cite{afgan_galaxy_2018} (first released 2005), Code~Ocean (2017), brainlife.io\cite{hayashi_brainlifeio_2024}, a non-exhaustive list, provided turnkey reproducibility services, each implementing separation of code, data, and environment as a first-class abstraction---reflecting Self-containment~(S), Actionability~(A), and Portability~(P).

Reviewer: ```suggestion
Though not an exhaustive list: Galaxy\cite{afgan_galaxy_2018} (first released 2005), Code~Ocean (2017), and brainlife.io\cite{hayashi_brainlifeio_2024} each provided turnkey reproducibility services, each implementing separation of code, data, and environment as a first-class abstraction---reflecting Self-containment~(S), Actionability~(A), and Portability~(P).
```

The BrainLife reference is giving our neuroscience / neuroimaging bias away, hah. There are no doubt many others we do not know of outside our field.

Also, DataJoint now does a lot of work with pharmacy so not just  neuroscience anymore, maybe work including them there too

Author: we could indeed. I do not mind promoting DataJoint.

On the standards side, the W3C~PROV data model (2013), Frictionless Data (2007), RO-Crate\cite{peroni_packaging_2022}, and BioComputeObject (IEEE~2791-2020) each addressed interoperable packaging and provenance recording.
The facts that FAIR required elaboration to accommodate software\cite{barker_introducing_2022} and workflows\cite{goble_fair_2020,wilkinson_applying_2025}, and that independent standards converged on provenance and packaging, confirms that these dimensions are recurrently demanded by research practice.

Framework tooling and domain standards reveal the same pattern at the project-layout level.

> Cookiecutter Data Science (2016), BIDS\cite{gorgolewski_brain_2016} (2016), Kedro (2019), nipoppy (2023), and others, all imposed opinionated directory structures and modular separation---independently motivated by the costs of ad-hoc layouts---reflecting Modularity~(M) and Self-containment~(S).

Reviewer: I would toss in PENNsieve into this mix: https://www.nature.com/articles/s41597-025-06075-5

Maybe DataJoint / AIND as well

But we are make the neuroscience bias clear yet again. Need to find and refer to some outside the field if at all possible

Author: Dont have enough experience with PENNsieve to assess its appropriateness here. Given that it is not even open source -- do not have strong desire to cite here.

DataJoint, which I would not mind promoting -- I do see it as a study pipeline, but given that it is mostly oriented around database for actual data, I do not see it fitting well this paragraph accenting on directory structures and "modular separation".  If you see how -- please suggest wording

Reviewer: >  Given that it is not even open source

It is open source: https://github.com/Pennsieve

> DataJoint, which I would not mind promoting -- I do see it as a study pipeline, but given that it is mostly oriented around database for actual data, I do not see it fitting well this paragraph accenting on directory structures and "modular separation".

They use it, especially in conjuction with SpyGlass, to promote as a way of doing all your science in a reproducible way with provenance, so very study-like by design. They are modular in that they have different elements per data stream for your experiment.

Reviewer: offloaded to #114 so you can proceed

> Each framework constrains where raw data, derived outputs, code, and documentation reside, making projects navigable by \emph{convention} rather than documentation alone.

Reviewer: ```suggestion
Each framework constrains where code and documentation reside, how raw data inputs and derived outputs are handled, which makes such projects navigable by \emph{convention}.
```

Some rewording

Viewed together, these independent lines of development map onto every STAMPED property.
Self-containment~(S) appears wherever projects bundle their dependencies into a single rooted structure; 
Tracking~(T) wherever content-addressed storage or version control records the history of changes or acquisition of every component;
Actionability~(A) wherever Makefiles, workflow managers, or platform APIs make re-execution or acquisition a single command;
Modularity~(M) wherever directory conventions or subproject composition separate concerns;
Portability~(P) wherever good coding practice, containers or environment specifications decouple a research object from a particular machine;

> Ephemerality~(E) wherever computational compositions bringing together compute and data are treated as disposable and regenerable;

Reviewer: ```suggestion
Ephemerality~(E) wherever composition brings together computation on data which is treated as disposable and regenerable;
```

Suggested rewording, initial reading was a bit scrambling to my mind

> and Distributability~(D) wherever licensing, packaging standards, or federated storage enable sharing.

Reviewer: One quick question Yarik, was this sentence AI-generated? If so I find it really loves to use that word 'federated'. We may want to include a reduction of that in the follow-up scan

Reviewer 2: +1 for a humanization pass:
 - https://github.com/stamped-principles/stamped-paper/issues/111

Author: let's go digging:

```
❯ git blame enh-lit -- main.tex | grep federated
a6059e12 (CodyCBakerPhD      2026-03-06 13:25:14 -0500 413) The same pattern---flexibility versus simplicity---recurs across other tool choices in Table~\ref{tab:stamped-tools}: container-based versus package-based environment management, workflow engines versus shell scripts, and federated archives versus centralized platforms.
30cc14cb (Yaroslav Halchenko 2026-03-21 14:20:37 -0400 506) Despite differing architectures---some cloud-native, some local-first, some centralized, some federated---each converged on the same core insight: that version-control semantics proven for code could and should extend to data.
30cc14cb (Yaroslav Halchenko 2026-03-21 14:20:37 -0400 525) and Distributability~(D) wherever licensing, packaging standards, or federated storage enable sharing.
```

```
❯ git show 30cc14cb | head -n 15
commit 30cc14cb0bace76298ad63cf8a3fc2eb30168fd9
Author: Yaroslav Halchenko <debian@onerussian.com>
Date:   Sat Mar 21 14:20:37 2026 -0400

    Rewrite convergent evolution section as temporal narrative
    
    Replace tool-by-tool subsubsections and non-STAMPED bullet lists with
    seven chronological paragraphs tracing how independent communities
    converged on patterns STAMPED formalizes.  Resolve intro FAIR TODO
    with STAMPED-vs-FAIR positioning.  Add foundational references
    (Claerbout 1992, Buckheit/Donoho 1995, Gentleman 2007, Stodden 2010,
    Sandve 2013, Galaxy 2018, brainlife.io 2024, RO-Crate 2022).
    
    Co-Authored-By: Claude Code 2.1.81 / Claude Opus 4.6 <noreply@anthropic.com>
```

so here a good chance that indeed this was AI produced, although federation is one of the key tenets of DataLad, see e.g. good old https://docs.datalad.org/en/maint/related.html but I would have not said "federated storage", I would have 

```suggestion
and Distributability~(D) wherever licensing, packaging standards, or data logistic mechanisms enable sharing.
```

or alike to accent that at times there simply no mechanisms provisioned in data management layer which would allow for sharing. WDYT?

Author: since it seems to be the only last one left -- I will just accept along with other suggestions so we could just merge and go from there

The formalization presented here provides a shared vocabulary for what these practices have in common, enabling researchers to evaluate and incrementally improve their workflows regardless of the specific tools they employ.
Figure~\ref{fig:convergent-evolution} summarizes these parallel timelines and their mapping to STAMPED properties.

\begin{figure}[htbp]
\centering
\includegraphics[width=\textwidth]{figures/convergent-evolution-gantt.pdf}
\caption{Convergent evolution of empirical scientific practice toward STAMPED.
Five independent lines of development---foundational ideas, normative frameworks, data version-control tools, platforms and metadata standards, and framework tooling---are shown on a shared timeline from 1992 to 2025.
Each milestone is annotated with the STAMPED properties it anticipates.

> Despite independent origins, all five tracks converged on the same organizational patterns that STAMPED formalizes.}

Reviewer: Would add some mention again here about the figure values not being exhaustive. Someone might just scan directly to the figures/tables and scoff if they don't see something common from their own field or an apologetic mention of unintentional omission

Reviewer 2: I added a sentence for this

These adoptions demonstrate that STAMPED properties solve practical organizational challenges across scales, from individual studies to community-wide infrastructure.

\subsection{Convergent evolution of data-driven scientific practices}

The challenges that motivate STAMPED are not unique to any single domain.

> Over three decades, communities spanning geophysics, statistics, genomics, neuroimaging, and machine learning independently converged on strikingly similar organizational responses to the same recurring problems: fragmented provenance, irreproducible computations, and ad-hoc project layouts.

Reviewer 2: why three decades? does this issue actually have a beginning? or did it just evolve since the beginning of science

Author: It is "Over" the three decades so clearly more and we are not arguing to know when it started, and **three** as that is where we started to care in our figure.
